src/unit.ts

```diff
@@ -0,0 +1,34 @@
+/**
+ * There can be only one.
+ *
+ * Useful for when you want a type of which there is guaranteed to ever only be
+ * one value. Examples include: indicating success or using it as a sentinel.
+ * null also fills this role, except you cannot refer to the Null type.
+ */
+class Unit {
+    // Public for tests.
+    static _instance: Unit;
+
+    /**
+     * instance returns the single Unit value.
+     */
+    public static instance(): Unit {
+        return this._instance || new Unit();
+    }
+
+    /**
+     * While Unit's constructor will still guarantee that it will not create
+     * duplicate values, you should prefer to call Unit.instance() instead,
+     * since its name captures the semantics of this behavior better.
+     */
+    constructor() {
+        if (typeof Unit._instance !== "undefined" && Unit._instance !== null) {
+            return Unit._instance;
+        } else {
+            Unit._instance = this;
+            return this;
+        }
+    }
+}
+
+export = Unit;
```

Comment:
Actually, being too fancy. I don't believe the `class` adds any extra type information, because structural typing. This is just as good:
```ts
interface Unit {
    __meow__: {};
}
var Unit = {
    unit: { __meow: {} }
};
export = Unit;
```